OrchardCore.Commerce/ViewModels/ShoppingCartLineViewModel.cs

```diff
@@ -1,18 +1,31 @@
-using System.Collections.Generic;
 using Money;
 using OrchardCore.Commerce.Abstractions;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OrchardCore.Commerce.ViewModels;
 
-namespace OrchardCore.Commerce.ViewModels
+public class ShoppingCartLineViewModel
 {
-    public class ShoppingCartLineViewModel
-    {
-        public int Quantity { get; set; }
-        public string ProductSku { get; set; }
-        public string ProductName { get; set; }
-        public string ProductUrl { get; set; }
-        public string ProductImageUrl { get; set; }
-        public Amount UnitPrice { get; set; }
-        public Amount LinePrice { get; set; }
-        public IDictionary<string, IProductAttributeValue> Attributes { get; set; }
-    }
-}
\ No newline at end of file
+    public int Quantity { get; set; }
+    public string ProductSku { get; set; }
+    public string ProductName { get; set; }
+    public string ProductUrl { get; set; }
+    public string ProductImageUrl { get; set; }
+    public Amount UnitPrice { get; set; }
+    public Amount LinePrice { get; set; }
+    public IDictionary<string, IProductAttributeValue> Attributes { get; }
+
+    public ShoppingCartLineViewModel(IDictionary<string, IProductAttributeValue> attributes = null) =>
+        Attributes = attributes ?? new Dictionary<string, IProductAttributeValue>();
+
+    public static bool IsSameProductAs(ShoppingCartLineViewModel line, ShoppingCartLineViewModel other) =>
+        other.ProductSku == line.ProductSku &&
+        (
+            (
+                (line.Attributes is null || line.Attributes.Count == 0) &&
+                (other.Attributes is null || other.Attributes.Count == 0)
+            ) ||
+            (line.Attributes?.Count == other.Attributes.Count && !line.Attributes.Except(other.Attributes).Any())
+        );
```

Comment:
It's not clear to me why you omit the `line.Attributes?.Count == other.Attributes.Count`? Maybe I'm misunderstanding but I think your version will give a false positive if `line` has more items than `other`.

On the other hand
```csharp
            (
                (line.Attributes is null || line.Attributes.Count == 0) &&
                (other.Attributes is null || other.Attributes.Count == 0)
            )
```
is 100% superfluous because  `line.Attributes?.Count` would just return `0` so the next line checks that and more.

I went with this, hope it's okay.
```csharp
    public static bool IsSameProductAs(ShoppingCartLineViewModel line, ShoppingCartLineViewModel other) =>
        other.ProductSku == line.ProductSku &&
        line.Attributes?.Count == other.Attributes.Count &&
        !line.Attributes.Except(other.Attributes).Any();
```

Reply:
Yes, you're right, the `Count` comparison is necessary. Looks good (except for the `?`, but you've already removed that by now, I think).